src/Core/Context/LegacyControllerContextBuilder.php

```diff
@@ -0,0 +1,224 @@
+<?php
+/**
+ * Copyright since 2007 PrestaShop SA and Contributors
+ * PrestaShop is an International Registered Trademark & Property of PrestaShop SA
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.md.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Core\Context;
+
+use Doctrine\ORM\NoResultException;
+use PrestaShop\PrestaShop\Adapter\ContextStateManager;
+use PrestaShop\PrestaShop\Adapter\Feature\MultistoreFeature;
+use PrestaShop\PrestaShop\Core\Domain\Shop\ValueObject\ShopConstraint;
+use PrestaShop\PrestaShop\Core\Util\Inflector;
+use PrestaShopBundle\Entity\Repository\TabRepository;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Contracts\Translation\TranslatorInterface;
+use Tools;
+
+class LegacyControllerContextBuilder implements LegacyContextBuilderInterface
+{
+    public function __construct(
+        private readonly EmployeeContext $employeeContext,
+        private readonly ShopContext $shopContext,
+        private readonly ContextStateManager $contextStateManager,
+        private readonly RequestStack $requestStack,
+        private readonly array $controllersLockedToAllShopContext,
+        private readonly TabRepository $tabRepository,
+        private readonly TranslatorInterface $translator,
+        private readonly MultistoreFeature $multistoreFeature,
+        private readonly ContainerInterface $container,
+        private readonly string $coreDir,
+    ) {
+    }
+
+    public function buildLegacyContext(): void
+    {
+        $request = $this->requestStack->getMainRequest();
```

Comment:
I'm never a big fan about injecting the request stack directly (traumatism from a previous project where we abused this, and it quickly turned into an anti-pattern that caused unexpected side effects 😅).

More generally in the case of these builders the question is about what **should** be handled internally and independently by the builder (not what could because it's always possible) and what should be passed via setters by the listeners that do have the request and can extract the information. The setter approach offers more reusability as it can be reused in use cases where we don't have a request for example.

Reply:
TLDR; maybe the controller name should be set via a listener instead? Although it's debatable because the current implementation doesn't need a listener and works out of the box simply by implementing the `LegacyContextBuilderInterface` so it's worth considering both possibilities